src/core/annotation.js

```diff
@@ -1516,6 +1603,23 @@ class TextWidgetAnnotation extends WidgetAnnotation {
 
     return chunks;
   }
+
+  getAnnotationObject() {
```

Comment:
Why do we need this `getAnnotationObject` method? The data passed between the core and display layers is all in the `this.data` attribute, and most of that information is already in there, so why can't we just use that?

Reply:
Because in this case we'll get only the data for the rendered Fields which should be fine most of the time.
But if a script has something like `var foo = this.getField("foo");do_whatever_with(foo);` and if foo hasn't been rendered then foo will be undefined. So in this case we could get lazily foo but it means that we should wait for an async function.
So the only solution I found (which doesn't mean that's the only solution) is to get all Fields.
To avoid to much redundancy, we could add a cache (if none) for the annotations (at least for the Fields): I guess we could expect that a user will read all the form just to be sure that no data are missing, it'll add some overhead at document loading but annotation rendering will be then slightly faster since we've a cache (but we use more memory...).
If you've any ideas on how to do that nicely then feel free to share them, I'm a kinda stuck here.

Reply:
Right, this explanation does help to get a bit more context here. Do we actually have examples of PDF files that do that or is it more a theoretical problem (is it even allowed by the spec to access fields from other pages)? If you really need all fields, I suppose that fetching the information from the `AcroForm` dictionary is the best as it doesn't require rendering all pages, which I think would be a no-go. I'm not sure though if the `AcroForm` dictionary is guaranteed to contain all form fields (some PDFs don't have the `AcroForm` dictionary, or work with `XFA`)...

Reply:
For me this still leaves the question why this method is necessary instead of just using `this.data` because in `document.js` you're using `AnnotationFactory.create` which will still create the entire annotation object anyway. Instead of calling `annotation.getAnnotationObject()` you might as well use `annotation.data` as far as I can tell. This would at least make sure we have one and only one way of exposing the data, and practically all fields in this method are already in `this.data` anyway...

Reply:
> Right, this explanation does help to get a bit more context here. Do we actually have examples of PDF files that do that or is it more a theoretical problem (is it even allowed by the spec to access fields from other pages)?

I don't have such a pdf but I guess it's pretty rare but having some fields to fill on page 1 and a field with the total sum on page 2 is likely existing (at least in France, forms for taxes are pretty long).
The method `Doc.getField` (https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/js_api_reference.pdf#page=301) allows to get a `Field` by its name (see page 413 too) and I don't see any mention that the `Field` must have been rendered.

> If you really need all fields, I suppose that fetching the information from the `AcroForm` dictionary is the best as it doesn't require rendering all pages, which I think would be a no-go. I'm not sure though if the `AcroForm` dictionary is guaranteed to contain all form fields (some PDFs don't have the `AcroForm` dictionary, or work with `XFA`)...

I don't know too. It should possible to build `AcroForm::Fields` using all the found annotations and their `T` field but in this case it'd require to read all the pages...
And about XFA, we plane to do that later.

Reply:
> For me this still leaves the question why this method is necessary instead of just using `this.data` because in `document.js` you're using `AnnotationFactory.create` which will still create the entire annotation object anyway. Instead of calling `annotation.getAnnotationObject()` you might as well use `annotation.data` as far as I can tell. This would at least make sure we have one and only one way of exposing the data, and practically all fields in this method are already in `this.data` anyway...

Agreed, I thought doing that but to have correct names it'd have required more change and I wanted to limit the changes in this patch. And by the way they're some "useless" (because they aren't in `Field` object) props in data.
I think we can do that in a follow-up (even in this one if we've an agreement on the way to do). 
An idea is to put these data (with correct names) in `AnnotationStorage` since I use this storage to store data coming from JS in case the annotation hasn't been rendered. Wdyt ?

Reply:
After thinking again about that, we can change names in `annotation.data` but it means that we potentially break something (api compat) and I'll still need to strip out useless properties at some point.

Reply:
You're right that we should break the API compatibility. How about instead of changing anything in `this.data` just adding a new key with the data you need for the JavaScript actions? In that case you can just augment `this.data` with a new key, say `javaScriptActions` with has an object as value, put your key/value pairs in there and when fetching the data only taking that key and ignoring the existing fields.

Reply:
I can do that but it means that we'll keep a ref on these data and once they have been consumed by the sandbox we don't need them anymore. So we'll consume memory for nothing.
But if your purpose is to expose that stuff, I'm fine to change that of course.

Reply:
See the new comment below. I only now understood why you actually need this.